## Content

- **Introduction to use case**

    
- **2-D arrays (Matrices)**
    - `reshape()`
    - Transpose
    - Converting Matrix back to Vector - `flatten()`


- **Indexing and Slicing on 2D** 
    - Indexing
    - Slicing
    - Masking (Fancy Indexing)

- **Universal Functions (ufunc) on 2D**
    - Aggregate Function/ Reduction functions - `sum()`, `mean()`, `min()`, `max()`
    - Axis argument

    - Logical Operations
    
    - Sorting function - `sort()`, `argsort()`
       
- **Use Case: Fitness Data analysis**
    - Loading data set and EDA using numpy
    - `np.argmax()`




## Use Case: Fitbit

### Imagine you are a Data Scientist at Fitbit

You've been given a user data to analyse and find some insights which can be shown on the smart watch.

#### But why would we want to analyse the user data for desiging the watch?

These insights from the user data can help business make customer oriented decision for the product design.



#### Lets first look at the data we have gathered

Link: https://drive.google.com/file/d/1Uxwd4H-tfM64giRS1VExMpQXKtBBtuP0/view?usp=sharing

<img src='https://drive.google.com/uc?id=1Uxwd4H-tfM64giRS1VExMpQXKtBBtuP0'>


#### Notice that there are some user features in the data

There are provided as various columns in the data.

#### Every row is called a record or data point


#### What are all the features provided to us? 

- Date
- Step Count
- Mood (Categorical)
- Calories Burned
- Hours of sleep
- Feeling Active (Categorical)


**Using NumPy, we will explore this data to look for some interesting insights - Exploratory Data Analysis.**

#### EDA is all about asking the right questions

#### What kind of questions can we answer using this data?

- How many records and features are there in the dataset?
- What is the **average step count**?
- On which day the **step count was highest/lowest?** 


#### Can we find some deeper insights?

We can probably see how daily activity affects sleep and moood.

We will try finding 
- How daily activity affects mood? 

In [ ]:
import numpy as np

## Working with 2-D arrays (Matrices)


#### Question : How do we create a matrix using numpy?

In [ ]:
m1 = np.array([[1,2,3],[4,5,6]])
m1
# Nicely printing out in a Matrix form

array([[1, 2, 3],
       [4, 5, 6]])

How can we check shape of a numpy array?

In [ ]:
m1.shape # arr1 has 3 elements

(2, 3)

#### Question: What is the type of this result of `arr1.shape`? Which data structure is this?
Tuple

#### Now, What is the dimension of this array?

In [ ]:
m1.ndim

2

#### Question
```

a = np.array([[1,2,3],
              [4,5,6],
              [7,8,9]])

b = len(a)
```
What'll be the value of b? 

```
Ans: 3
```
**Explanation: len(nD array) will give you magnitude of first dimension**


In [ ]:
a = np.array([[1,2,3],
              [4,5,6],
              [7,8,9]])

In [ ]:
a

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [ ]:
len(a)

3

#### What will be the shape of array `a`?

In [ ]:
a.shape

(3, 3)

- So, it is a **2-D array** with **3 rows and 3 columns**

Clearly, if **we have to create high-dimensional arrays, we cannot do this using `np.arange()`** directly

### How can we create high dimensional arrays?

- Using `reshape()`

For a 2D array
- **First argument** is **no. of rows**
- **Second argument** is **no. of columns**

In [ ]:
m2 = np.arange(1, 13)
m2

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

- We can pass the **desired dimensions** of array in `reshape()`

#### In what ways can we convert this array with 12 values into high-dimensional array?

#### Can we make `m2` a $4\times4$ array?

- Obviously NO
- **$4\times4$ requires 16 values**, but **we only have 12 in `m2`**

In [ ]:
m2 = np.arange(1, 13)
m2.reshape(4, 4)

ValueError: ignored

#### So, What are the ways in which we can reshape it?

- $4\times3$
- $3\times4$
- $6\times2$
- $2\times6$
- $1\times12$
- $12\times1$

In [ ]:
m2 = np.arange(1, 13)
m2.reshape(4, 3)

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [ ]:
m2 = np.arange(1, 13)
m2

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

In [ ]:
m2.shape

(12,)

#### Lets do some reshaping here 

In [ ]:
m2.reshape(12, 1)

array([[ 1],
       [ 2],
       [ 3],
       [ 4],
       [ 5],
       [ 6],
       [ 7],
       [ 8],
       [ 9],
       [10],
       [11],
       [12]])

#### Now, What's the difference b/w `(12,)` and `(12, 1)`?

- **`(12,)`** means its a **1D array**
- **`(12, 1)`** means its a **2D array**

#### Question

What will be output for the following code?
```
a = np.array([[1,2,3],[0,1,4]])
print(a.ndim)
```

**Ans: 2**


In [ ]:
a = np.array([[1,2,3],[0,1,4]])
print(a.ndim)

2


Since it is a 2 dimensional array, the number of dimension will be 2.

### Transpose

- **Change rows into columns and columns into rows**

- Just use `<Matrix>.T`

In [ ]:
a = np.arange(3)
a

array([0, 1, 2])

In [ ]:
a.T

array([0, 1, 2])

#### Why did Transpose did not work?

- Because **numpy sees `a` as a vector (3,), NOT a matrix**

- We'll have to **reshape the vector `a` to make it a matrix**

In [ ]:
a = np.arange(3).reshape(1, 3)
a
# Now a has dimensions (1, 3) instead of just (3,)
# It has 1 row and 3 columns

array([[0, 1, 2]])

In [ ]:
a.T
# It has 3 rows and 1 column

array([[0],
       [1],
       [2]])



#### Conclusion

- **Transpose works only on matrices**


### Flattening of an array

#### What if we want to convert this 2D or nD array back to 1D array?

There is a function named `flatten()` to help you do so.



In [ ]:
A = np.arange(12).reshape(3, 4)
A

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [ ]:
A.flatten() 

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

## Indexing and Slicing on 2D Numpy arrays





### Indexing in np arrays

- Works same as lists

In [ ]:
m1 = np.arange(1,10).reshape((3,3))

In [ ]:
m1

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [ ]:
m1[1][2]

6

OR

- We just use [0, 0] (**indexes separated by commas**)

#### What will be the output of this?

In [ ]:
m1[1, 1] #m1[row, column]

5

#### We saw how we can use list of indexes in numpy array

In [ ]:
m1 = np.array([100,200,300,400,500,600])

In [ ]:
m1[[2,3,4,1,2,2]]

array([300, 400, 500, 200, 300, 300])



How'll list of indexes work in 2D array ?

In [ ]:
m1 = np.arange(9).reshape((3,3))

In [ ]:
m1

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

In [ ]:
m1[[0,1,2],[0,1,2]] # picking up element (0,0), (1,1) and (2,2)

array([0, 4, 8])

### Slicing

- Need to **provide two slice ranges** - one for **row** and one for **column**
- Can also **mix Indexing and Slicing**

In [ ]:
m1 = np.arange(12).reshape(3,4)
m1

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [ ]:
m1[:2] # gives first two rows

array([[0, 1, 2, 3],
       [4, 5, 6, 7]])

#### How can we get columns from 2D array?

In [ ]:
m1[:, :2] # gives first two columns

array([[0, 1],
       [4, 5],
       [8, 9]])

**Question: Given an 2-D array**

```
m1 = [[0,1,2,3],
     [4,5,6,7],
     [8,9,10,11]]

```

In [ ]:
m1 = m1.reshape((3,4))

#### Question for you: Can you just get this much of our array `m1`?
```
[[5, 6],
 [9, 10]]
```

#### Remember our `m1` is:
```
m1 = [[0, 1, 2, 3],
      [4, 5, 6, 7],
      [8, 9, 10, 11]]
```

In [ ]:
# First get rows 1 to all
# Then get columns 1 to 3 (not included)
m1[1:, 1:3]

array([[ 5,  6],
       [ 9, 10]])

#### Question: What if I need 1st and 3rd column?
```
[[1, 3],
 [5, 7],
 [9,11]]
```

In [ ]:
# Get all rows
# Then get columns from 1 to all with step of 2

m1[:, 1::2]

array([[ 1,  3],
       [ 5,  7],
       [ 9, 11]])

- **We can also pass indices of required columns as a Tuple** to get the same result

In [ ]:
# Get all rows
# Then get columns 1 and 3

m1[:, (1,3)]

array([[ 1,  3],
       [ 5,  7],
       [ 9, 11]])

### Fancy indexing (Masking)
\


#### What would happen if we do this?


In [ ]:
m1 = np.arange(12).reshape(3, 4)
m1 < 6

array([[ True,  True,  True,  True],
       [ True,  True, False, False],
       [False, False, False, False]])


- A **matrix having boolean values** `True` and `False` is returned 



- **We can use this boolean matrix to filter our array**


#### Now, Let's use this to filter or mask values from our array

- **Condition will be passed instead of indices and slice ranges**

In [ ]:
m1[m1 < 6]
# Value corresponding to True is retained
# Value corresponding to False is filtered out

array([0, 1, 2, 3, 4, 5])




#### How can we filter/mask even values from our array?

In [ ]:
m1[m1%2 == 0]

array([ 0,  2,  4,  6,  8, 10])

#### But did you notice that matrix gets converted into a 1D array after masking?

In [ ]:
m1

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [ ]:
m1[m1%2 == 0]

array([ 0,  2,  4,  6,  8, 10])

#### It happens because

- To retain matrix shape, it **has to retain all the elements**
- It **cannot retain its $3\times4$ with lesser number of elements**
- So, this filtering operation **implicitly converts high-dimensional array into 1D array**


#### If we want, we can reshape the resulting 1D array into 2D

- But, we need to know **beforehand** what is the **dimension or number of elements** in resulting 1D array

In [ ]:
m1[m1%2==0].shape

(6,)

In [ ]:
m1[m1%2==0].reshape(2, 3)

array([[ 0,  2,  4],
       [ 6,  8, 10]])

## Universal Functions (`ufunc`) on 2D & Axis 

### Aggregate Functions/ Reduction functions

We saw how aggregate functions work on 1D array in last class



In [ ]:
arr = np.arange(3)
arr

array([0, 1, 2])

In [ ]:
arr.sum()

3

#### Let's apply Aggregate functions on 2D array

### `np.sum()`

In [ ]:
a = np.arange(12).reshape(3, 4)
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [ ]:
np.sum(a)  # sums all the values present in array

66

#### What if we want to do the elements row-wise or column-wise?

- By **setting `axis` parameter**

#### What will `np.sum(a, axis=0)` do?

- **`np.sum(a, axis=0)` adds together values in DIFFERENT rows**
- **`axis = 0` ---> Changes will happen along the vertical axis**
- Summing of values happen **in the vertical direction**
- Rows collapse/merge when we do `axis=0`


In [ ]:
np.sum(a, axis=0)

array([12, 15, 18, 21])

#### Now, What if we specify `axis=1`?

- **`np.sum(a, axis=1)` adds together values in DIFFERENT columns** 
- **`axis = 1` ---> Changes will happen along the horizontal axis**
- Summing of values happen **in the horizontal direction**
- Columns collapse/merge when we do `axis=1`

In [ ]:
np.sum(a, axis=1)

array([ 6, 22, 38])

***

#### Now, What if we want to find the average value or median value of all the elements in an array?

In [ ]:
np.mean(a) # no need to give any axis

5.5

#### What if we want to find the mean of elements in each row or in each column?

- We can do **same thing with `axis` parameter** like we did for `np.sum()` function


#### Question: Now you tell What will `np.mean(a, axis=0)` give?

- It will give **mean of values in DIFFERENT rows**
- **`axis = 0` ---> Changes will happen along the vertical axis**
- Mean of values will be calculated **in the vertical direction**
- Rows collapse/merge when we do `axis=0`

In [ ]:
np.mean(a, axis=0)

array([4., 5., 6., 7.])

#### How can we get mean of elements in each column?

- **`np.mean(a, axis=1)` will give mean of values in DIFFERENT columns** 
- **`axis = 1` ---> Changes will happen along the horizontal axis**
- Mean of values will be calculated **in the horizontal direction**
- Columns collapse/merge when we do `axis=1`

In [ ]:
np.mean(a, axis=1)

array([1.5, 5.5, 9.5])

#### Now, we want to find the minimum value in the array

**`np.min()` function can help us with this**

In [ ]:
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [ ]:
np.min(a)

0

#### What if we want to find row wise minimum value?

### Use `axis` argument!!

In [ ]:
np.min(a, axis = 1 )

array([0, 4, 8])

#### We can also find max elements in an array.

**`np.max()`** function will give us *maximum value in the array*

We can also use `axis` argument to find row wise/ column wise max.

In [ ]:
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [ ]:
np.max(a) # maximum value

11

In [ ]:
np.max(a, axis = 0) # column wise max

array([ 8,  9, 10, 11])

### Logical Operations

#### Now, What if we want to check whether "any" element of array follows a specific condition?

#### Let's say we have 2 arrays:

In [ ]:
a = np.array([1,2,3,4])
b = np.array([4,3,2,1])
a, b

(array([1, 2, 3, 4]), array([4, 3, 2, 1]))

#### Let's say we want to find out if any of the elements in array `a` is smaller than any of the corresponding elements in array `b`


#### `np.any()` can become handy here as well


- `any()` returns `True` if **any of the corresponding elements** in the argument arrays follow the **provided condition**.


In [ ]:
a = np.array([1,2,3,4])
b = np.array([4,3,2,1])
np.any(a<b) # Atleast 1 element in a < corresponding element in b

True

#### Let's try the same condition with different arrays:

In [ ]:
a = np.array([4,5,6,7])
b = np.array([4,3,2,1])
np.any(a<b) # All elements in a >= corresponding elements in b

False

- In this case, **NONE of the elements in `a` were smaller than their corresponding elements in `b`**

- So, `np.any(a<b)` returned `False`

***

#### What if we want to check whether "all" the elements in our array are non-zero or follow the specified condition?

`np.all()`


#### Now, What if we want to check whether "all" the elements in our array follow a specific condition?




#### Let's say we want to find out if all the elements in array `a` are smaller than all the corresponding elements in array `b`

Again, Let's say we have 2 arrays:


In [ ]:
a = np.array([1,2,3,4])
b = np.array([4,3,2,1])
a, b

(array([1, 2, 3, 4]), array([4, 3, 2, 1]))

In [ ]:
np.all(a<b) # Not all elements in a < corresponding elements in b

False

#### Let's try it with different arrays

In [ ]:
a = np.array([1,0,0,0])
b = np.array([4,3,2,1])
np.all(a<b) # All elements in a < corresponding elements in b

True

- In this case, **ALL the elements in `a` were smaller than their corresponding elements in `b`**

- So, `np.all(a<b)` returned `True`

#### Multiple conditions for `.all()` function

In [ ]:
a = np.array([1, 2, 3, 2])
b = np.array([2, 2, 3, 2])
c = np.array([6, 4, 4, 5])
((a <= b) & (b <= c)).all()

True

#### What if we want to update an array based on condition ? 

Suppose you are given an array of integers and you want to update it based on following condition:
- if element is > 0, change it to +1
- if element < 0, change it to -1.

#### How will you do it ? 


In [ ]:
arr = np.array([-3,4,27,34,-2, 0, -45,-11,4, 0 ])
arr

array([ -3,   4,  27,  34,  -2,   0, -45, -11,   4,   0])

You can use masking to update the array (as discussed in last class)

In [ ]:
arr[arr > 0]  = 1
arr [arr < 0] = -1

In [ ]:
arr

array([-1,  1,  1,  1, -1,  0, -1, -1,  1,  0])

There is a numpy function which can help us with it.

#### np.where()

Function signature: 
`np.where(condition, [x, y])`

This functions returns an ndarray whose elements are chosen from x or y depending on condition.



In [ ]:
arr = np.array([-3,4,27,34,-2, 0, -45,-11,4, 0 ])


In [ ]:
np.where(arr > 0, +1, -1)

array([-1,  1,  1,  1, -1, -1, -1, -1,  1, -1])

In [ ]:
arr

array([ -3,   4,  27,  34,  -2,   0, -45, -11,   4,   0])

Notice that it didn't change the original array.

### Sorting Arrays

- We can also sort the elements of an array along a given specified axis 


- Default axis is the last axis of the array.



#### `np.sort()`



In [ ]:
a = np.array([2,30,41,7,17,52])
a

array([ 2, 30, 41,  7, 17, 52])

In [ ]:
np.sort(a)

array([ 2,  7, 17, 30, 41, 52])

In [ ]:
a

array([ 2, 30, 41,  7, 17, 52])

Let's work with 2D array

In [ ]:
a = np.arange(9,0,-1).reshape(3,3)
a

array([[9, 8, 7],
       [6, 5, 4],
       [3, 2, 1]])

#### Question: What will be the result when we sort using axis = 0 ?

In [ ]:
np.sort(a, axis = 0)

array([[3, 2, 1],
       [6, 5, 4],
       [9, 8, 7]])

Recall that when axis =0
- change will happen along vertical axis.

Hence, it will sort out row wise.

In [ ]:
a

array([[9, 8, 7],
       [6, 5, 4],
       [3, 2, 1]])

- Original array is still the same. It hasn't changed

#### `np.argsort()`

- Returns the **indices** that would sort an array.

- Performs an indirect sort along the given axis. 

- It returns **an array of indices of the same shape as a that index data along the given axis in sorted order**.

In [ ]:
a = np.array([2,30,41,7,17,52])
a

array([ 2, 30, 41,  7, 17, 52])

In [ ]:
np.argsort(a)

array([0, 3, 4, 1, 2, 5])

#### As you can see:

- The orginal indices of elements are in same order as the orginal elements would be in sorted order

## Use Case: Fitness data analysis



### Let's first download the dataset

In [ ]:
!gdown 1vk1Pu0djiYcrdc85yUXZ_Rqq2oZNcohd


Downloading...
From: https://drive.google.com/uc?id=1vk1Pu0djiYcrdc85yUXZ_Rqq2oZNcohd
To: /content/fit.txt
100% 3.43k/3.43k [00:00<00:00, 4.10MB/s]


### Let's load the data we saw earlier. For this we will use `.loadtxt() function`

In [ ]:
data = np.loadtxt('fit.txt', dtype='str')

We provide file name along with the dtype of data we want to load in 

In [ ]:
data[:5]

array([['06-10-2017', '5464', 'Neutral', '181', '5', 'Inactive'],
       ['07-10-2017', '6041', 'Sad', '197', '8', 'Inactive'],
       ['08-10-2017', '25', 'Sad', '0', '5', 'Inactive'],
       ['09-10-2017', '5461', 'Sad', '174', '4', 'Inactive'],
       ['10-10-2017', '6915', 'Neutral', '223', '5', 'Active']],
      dtype='<U10')

What's the shape of the data? 

In [ ]:
data.shape

(96, 6)

There are 96 records and each record has 6 features.
These features are:
- Date
- Step count
- Mood
- Calories Burned
- Hours of sleep
- activity status

#### Notice that above array is a homogenous containing all the data as strings

In order to work with strings, categorical data and numerical data, we will have save every feature seperately

#### How will we extract features in seperate variables?

We can get some idea on how data is saved.

Lets see whats the first element of `data`

In [ ]:
data[0]

array(['06-10-2017', '5464', 'Neutral', '181', '5', 'Inactive'],
      dtype='<U10')

Hm, this extracts a row not a column

Think about it.

#### Whats the way to change columns to rows and rows to columns?

Transpose

In [ ]:
data.T[0]

array(['06-10-2017', '07-10-2017', '08-10-2017', '09-10-2017',
       '10-10-2017', '11-10-2017', '12-10-2017', '13-10-2017',
       '14-10-2017', '15-10-2017', '16-10-2017', '17-10-2017',
       '18-10-2017', '19-10-2017', '20-10-2017', '21-10-2017',
       '22-10-2017', '23-10-2017', '24-10-2017', '25-10-2017',
       '26-10-2017', '27-10-2017', '28-10-2017', '29-10-2017',
       '30-10-2017', '31-10-2017', '01-11-2017', '02-11-2017',
       '03-11-2017', '04-11-2017', '05-11-2017', '06-11-2017',
       '07-11-2017', '08-11-2017', '09-11-2017', '10-11-2017',
       '11-11-2017', '12-11-2017', '13-11-2017', '14-11-2017',
       '15-11-2017', '16-11-2017', '17-11-2017', '18-11-2017',
       '19-11-2017', '20-11-2017', '21-11-2017', '22-11-2017',
       '23-11-2017', '24-11-2017', '25-11-2017', '26-11-2017',
       '27-11-2017', '28-11-2017', '29-11-2017', '30-11-2017',
       '01-12-2017', '02-12-2017', '03-12-2017', '04-12-2017',
       '05-12-2017', '06-12-2017', '07-12-2017', '08-12

Great, we could extract first column

#### Lets extract all the columns and save them in seperate variables

In [ ]:
date, step_count, mood, calories_burned, hours_of_sleep, activity_status = data.T

In [ ]:
step_count

array(['5464', '6041', '25', '5461', '6915', '4545', '4340', '1230', '61',
       '1258', '3148', '4687', '4732', '3519', '1580', '2822', '181',
       '3158', '4383', '3881', '4037', '202', '292', '330', '2209',
       '4550', '4435', '4779', '1831', '2255', '539', '5464', '6041',
       '4068', '4683', '4033', '6314', '614', '3149', '4005', '4880',
       '4136', '705', '570', '269', '4275', '5999', '4421', '6930',
       '5195', '546', '493', '995', '1163', '6676', '3608', '774', '1421',
       '4064', '2725', '5934', '1867', '3721', '2374', '2909', '1648',
       '799', '7102', '3941', '7422', '437', '1231', '1696', '4921',
       '221', '6500', '3575', '4061', '651', '753', '518', '5537', '4108',
       '5376', '3066', '177', '36', '299', '1447', '2599', '702', '133',
       '153', '500', '2127', '2203'], dtype='<U10')

In [ ]:
step_count.dtype

dtype('<U10')

Notice the data type of step_count and other variables. It's a string type where **U** means Unicode String. and 10 means 10 bytes. 

#### Why? Because Numpy type-casted all the data to strings.

#### Let's convert the data types of these variables

**Step Count**

In [ ]:
step_count = np.array(step_count, dtype = 'int')
step_count.dtype

dtype('int64')

In [ ]:
step_count

array([5464, 6041,   25, 5461, 6915, 4545, 4340, 1230,   61, 1258, 3148,
       4687, 4732, 3519, 1580, 2822,  181, 3158, 4383, 3881, 4037,  202,
        292,  330, 2209, 4550, 4435, 4779, 1831, 2255,  539, 5464, 6041,
       4068, 4683, 4033, 6314,  614, 3149, 4005, 4880, 4136,  705,  570,
        269, 4275, 5999, 4421, 6930, 5195,  546,  493,  995, 1163, 6676,
       3608,  774, 1421, 4064, 2725, 5934, 1867, 3721, 2374, 2909, 1648,
        799, 7102, 3941, 7422,  437, 1231, 1696, 4921,  221, 6500, 3575,
       4061,  651,  753,  518, 5537, 4108, 5376, 3066,  177,   36,  299,
       1447, 2599,  702,  133,  153,  500, 2127, 2203])

**Calories Burned**

In [ ]:
calories_burned = np.array(calories_burned, dtype = 'int')
calories_burned.dtype

dtype('int64')

**Hours of Sleep**

In [ ]:
hours_of_sleep = np.array(hours_of_sleep, dtype = 'int')
hours_of_sleep.dtype

dtype('int64')

**Mood**

`Mood` is a categorical data type. As a name says, categorical data type has two or more categories in it.

Let's check the values of `mood` variable

In [ ]:
mood

array(['Neutral', 'Sad', 'Sad', 'Sad', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Sad', 'Sad', 'Sad', 'Sad', 'Happy', 'Sad', 'Sad', 'Sad', 'Sad',
       'Neutral', 'Neutral', 'Neutral', 'Neutral', 'Neutral', 'Neutral',
       'Happy', 'Neutral', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Neutral', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Neutral',
       'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Neutral', 'Sad', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Happy', 'Happy', 'Sad', 'Neutral', 'Neutral',
       'Sad', 'Sad', 'Neutral', 'Neutral', 'Happy', 'Neutral', 'Neutral',
       'Sad', 'Neutral', 'Sad', 'Neutral', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Sad', 'Happy', 'Neutral', 'Happy', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Neutral', 'Neutral', 'Sad', 'Sad', 'Happy', 'Neutral', 'Neutral',
       'Happy'], dtype='<U10')

In [ ]:
np.unique(mood)

array(['Happy', 'Neutral', 'Sad'], dtype='<U10')

**Activity Status**

In [ ]:
activity_status

array(['Inactive', 'Inactive', 'Inactive', 'Inactive', 'Active',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Active', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Active', 'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Active',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Active', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Inactive', 'Active', 'Active',
       'Inactive', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Inactive', 'Active', 'Active', 'Active', 'Active', 'Inactive',
       'Inactive', 'Inactive', 'Inacti

### Let's try to get some insights from the data.

#### What's the average step count? 

How can we calculate average? => `.mean()`

In [ ]:
step_count.mean()

2935.9375

User moves an average of 2900 steps a day.

#### On which day the step count was highest?

How will be find it? 

First we find the index of maximum step count and use that index to get the date.

How'll we find the index? =>  

Numpy provides a function `np.argmax()` which returns the index of maximum value element.

Similarly, we have a function `np.argmin()` which returns the index of minimum element.

In [ ]:
step_count.argmax()


69

Here 69 is the index of maximum step count element.

In [ ]:
date[step_count.argmax()]

'14-12-2017'

Let's check the calorie burnt on the day

In [ ]:
calories_burned[step_count.argmax()]

243

Not bad! 243 calories. Let's try to get the number of steps on that day as well

In [ ]:
step_count.max()

7422

7k steps!! Sports mode on!

#### Let's try to compare step counts on bad mood days and good mood days

Average step count on Sad mood days

In [ ]:
np.mean(step_count[mood == 'Sad'])

2103.0689655172414

In [ ]:
np.sort(step_count[mood == 'Sad'])

array([  25,   36,   61,  133,  177,  181,  221,  299,  518,  651,  702,
        753,  799, 1230, 1258, 1580, 1648, 1696, 2822, 3148, 3519, 3721,
       4061, 4340, 4545, 4687, 5461, 6041, 6676])

In [ ]:
np.std(step_count[mood == 'Sad'])

2021.2355035376254

Average step count on happy days

In [ ]:
np.mean(step_count[mood == 'Happy'])

3392.725

In [ ]:
np.sort(step_count[mood == 'Happy'])

array([ 153,  269,  330,  493,  539,  546,  614,  705,  774,  995, 1421,
       1831, 1867, 2203, 2255, 2725, 3149, 3608, 4005, 4033, 4064, 4068,
       4136, 4275, 4421, 4435, 4550, 4683, 4732, 4779, 4880, 5195, 5376,
       5464, 5537, 5934, 5999, 6314, 6930, 7422])

Average step count on sad days - 2103.

Average step count on happy days - 3392

There may be relation between mood and step count

#### Let's try to check inverse. Mood when step count was greater/lesser

Mood when step count > 4000

In [ ]:
np.unique(mood[step_count > 4000], return_counts = True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([22,  9,  7]))

Out of 38 days when step count was more than 4000, user was feeling happy on 22 days.

Mood when step count <= 2000

In [ ]:
np.unique(mood[step_count < 2000], return_counts = True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([13,  8, 18]))

Out of 39 days, when step count was less than 2000, user was feeling sad on 18 days.

#### **There may be a correlation between Mood and step count**